src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1615,26 +1611,32 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
 // is held, to ensure that the compiler is not using the class hierarchy, and that deoptimization will kick in
 // before a new class is used.
 
-void SystemDictionary::add_to_hierarchy(InstanceKlass* k) {
-  assert(k != nullptr, "just checking");
-  if (Universe::is_fully_initialized()) {
-    assert_locked_or_safepoint(Compile_lock);
-  }
+void SystemDictionary::add_to_hierarchy(InstanceKlass* k, JavaThread* current) {
+  DeoptimizationScope deopt_scope;
+  {
+    MutexLocker ml(current, Compile_lock);
+    assert(k != nullptr, "just checking");
+    if (Universe::is_fully_initialized()) {
+      assert_locked_or_safepoint(Compile_lock);
```

Comment:
This serves no purpose given you just took out the lock. But it does raise the question as to whether the lock can/should always be taken out if called at a safepoint, or if the Universe is not yet fully initialized?

Reply:
I found one call to `add_to_hierarchy()` where we previously did not
grab in `Compile_lock` in `vmClasses::resolve_shared_class()` and
that call is made with this assert in place:
```assert(!Universe::is_fully_initialized(), "We can make short cuts only during VM initialization");```

so it looks to me like we have a locking issue there (as in too early for locks).

Reply:
I think you are taking about CDS dump process. It works fine taking the lock.
I think this was just some micro-optimization. But we should not mark and do deopt fixed that.

Reply:
To restate: the assert_locked_or_safepoint is pointless because you just took out the lock here.

The code Dan points to does not seem to be related to CDS dump:
```
bool vmClasses::resolve(vmClassID id, TRAPS) {
  InstanceKlass** klassp = &_klasses[as_int(id)];

#if INCLUDE_CDS
  if (UseSharedSpaces && !JvmtiExport::should_post_class_prepare()) {
    InstanceKlass* k = *klassp;
    assert(k->is_shared_boot_class(), "must be");

    ClassLoaderData* loader_data = ClassLoaderData::the_null_class_loader_data();
    resolve_shared_class(k, loader_data, Handle(), CHECK_false);
    return true;
  }
#endif // INCLUDE_CDS
```
I remain concerned this code may be executed at a safepoint and we will now try to take the Compile_lock.

Reply:
I did not find in testing any case that executed the add_to_hierarchy.
I'll check more!